File: idakonyvtar/src/main/java/eu/eyan/idakonyvtar/view/MultiField.java
package eu.eyan.idakonyvtar.view;

import static com.google.common.collect.Lists.newArrayList;

import java.awt.Component;
import java.awt.event.ActionEvent;
import java.util.List;
import java.util.stream.Collectors;

import javax.swing.BoxLayout;
import javax.swing.JButton;
import javax.swing.JPanel;

import lombok.Getter;

import com.jgoodies.forms.builder.PanelBuilder;
import com.jgoodies.forms.factories.CC;
import com.jgoodies.forms.layout.FormLayout;

public abstract class MultiField<INPUT, EDITOR extends Component> extends JPanel implements FieldEditListener<EDITOR>
{
    private static final long serialVersionUID = 1L;
    private final List<Field<EDITOR>> fields = newArrayList();
    private String columnName;
    private int counter = 1;

    private static class Field<EDITOR>
    {
        @Getter
        private final EDITOR editor;
        @Getter
        private final JButton delete;
        @Getter
        private final JPanel panel;

        public Field(EDITOR editor, JButton button, JPanel mezőPanel)
        {
            this.editor = editor;
            this.delete = button;
            this.panel = mezőPanel;
        }
    }

    public MultiField(String columnName)
    {
        this.columnName = columnName;
        this.setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
    }

    public void setValues(List<INPUT> values)
    {
        removeAll();
        fields.clear();

        for (INPUT input : values)
        {
            addEditor(input, false);
        }
        addEditor(null, true);
    }

    private void addEditor(INPUT input, boolean last)
    {
        EDITOR editor = getEditor();
        addFieldEditListener(editor, this);
        JButton deleteButton = new JButton("x");

        PanelBuilder panelBuilder = new PanelBuilder(new FormLayout("f:p:g, 3dlu, 30dlu", "f:p:g, 3dlu"));
        panelBuilder.add(editor, CC.xy(1, 1));
        panelBuilder.add(deleteButton, CC.xy(3, 1));
        JPanel fieldPanel = panelBuilder.build();

        if (last)
        {
            deleteButton.setEnabled(false);
        }
        else
        {
            setValueInEditor(editor, input);
        }

        final Field<EDITOR> field = new Field<EDITOR>(editor, deleteButton, fieldPanel);
        deleteButton.addActionListener((ActionEvent actionEvent) -> {
            fields.remove(field);
            remove(field.getPanel());
            revalidate();
        });
        fields.add(field);
        add(fieldPanel);

        fieldPanel.setName(columnName + ".panel." + counter);
        editor.setName(columnName + counter);
        deleteButton.setName(columnName + ".delete." + counter);
        counter++;

        revalidate();
//        SwingUtilities.getWindowAncestor(this).pack();
    }

    protected abstract void addFieldEditListener(EDITOR editor, FieldEditListener<EDITOR> listener);

    @Override
    public void fieldEdited(EDITOR source)
    {
        Field<EDITOR> lastField = fields.get(fields.size() - 1);
        if (lastField.getEditor() == source)
        {
            lastField.getDelete().setEnabled(true);
            addEditor(null, true);
        }
    }

    protected abstract EDITOR getEditor();

    public List<INPUT> getValues()
    {
        return fields.stream().map((Field<EDITOR> field) -> {
            return getValue(field.getEditor());
        }).collect(Collectors.toList());
    }

    /**
     * @return null if empty!
     */
    protected abstract INPUT getValue(EDITOR editor);

    protected abstract void setValueInEditor(EDITOR editor, INPUT value);
}

File: idakonyvtar/src/main/scala/eu/eyan/idakonyvtar/view/MultiField.scala
package eu.eyan.idakonyvtar.view;

import java.awt.Component

import scala.collection.mutable.ListBuffer

import eu.eyan.log.Log
import eu.eyan.util.awt.ComponentPlus.ComponentPlusImplicit
import eu.eyan.util.swing.JButtonPlus.JButtonImplicit
import eu.eyan.util.swing.JPanelWithFrameLayout
import javax.swing.BoxLayout
import javax.swing.JButton
import javax.swing.JPanel
import eu.eyan.util.swing.JButtonPlus.JButtonImplicit

abstract class MultiField[INPUT, EDITOR <: Component](columnName: String) extends JPanel {

  def getValues: List[INPUT] = editors.map(_.editor).map(getValue).toList.filter(_.nonEmpty).map(_.get)

  def setValues(values: Array[INPUT]) = {
    Log.debug(getName + ": " + values.mkString(", "))
    removeAll
    editors.clear

    values foreach addEditorWithValue
    addEditorEmpty
  }

  protected def createEditor(fieldEdited: EDITOR => Unit): EDITOR
  protected def getValue(editor: EDITOR): Option[INPUT]
  protected def setValueInEditor(editor: EDITOR)(value: INPUT): Unit

  setLayout(new BoxLayout(this, BoxLayout.Y_AXIS))
  private val editors: ListBuffer[Editor[EDITOR]] = ListBuffer()
  private var counter = 0

  private case class Editor[EDITOR](val editor: EDITOR, val deleteButton: JButton)

  private def addEditorEmpty = addEditor(None)
  private def addEditorWithValue(input: INPUT) = addEditor(Some(input))

  private def addEditor(input: Option[INPUT]): Unit = {
    counter = counter + 1
    val editor = createEditor(addEditorIfLast).name(columnName + counter)
    input.foreach( setValueInEditor(editor) )

    val fieldPanel = new JPanelWithFrameLayout()
      .withSeparators
      .newColumn("f:p:g")
      .newRow("f:p:g")
      .name(columnName + ".panel." + counter)
      .addFluent(editor)
      .newColumn("30dlu")

    val deleteButton = fieldPanel
      .addButton("x")
      .name(columnName + ".delete." + counter)
      .enabled(input.nonEmpty)
      .onAction(removeEditor(editor))

    editors += new Editor[EDITOR](editor, deleteButton)
    add(fieldPanel)

    Log.debug(columnName + counter)
  }

  private def removeEditor(editor: EDITOR) = {
    editors.find(_.editor == editor).foreach(field => editors -= field)
    remove(editor.getParent)
  }

  private def addEditorIfLast(editor: EDITOR) =
    if (editors.last.editor == editor) {
      editors.last.deleteButton.enabled
      addEditorEmpty
    }
}